R/toolISOhistorical.R

```diff
@@ -115,56 +116,253 @@ toolISOhistorical <- function(m, mapping = NULL, additional_mapping = NULL, over
+                                                                     paste, sep = "."))),
+                                          c("fromISO", "toISO")],
+                             weight = NULL)
       }
 
-      mTr <- toolAggregate(m[a$fromISO, subTime, ],
-                           mapping[is.element(mapping$toISO, a$toISO), c("fromISO", "toISO")], weight = weight,
-                           negative_weight = "allow")
-      ## aggregation of countries
-    } else {
-      mTr <- toolAggregate(m[a$fromISO, subTime, ],
-                           mapping[is.element(mapping$toISO, a$toISO), c("fromISO", "toISO")], weight = NULL)
+      return(zTr)
     }
-    # fill data
-    if (overwrite == TRUE) {
-      m[a$toISO, subTime, ] <- mTr
-    } else {
-      m[a$toISO, subTime, ][is.na(m[a$toISO, subTime, ])] <- mTr[is.na(m[a$toISO, subTime, ])]
+
+    # loop over all transitions in dim 1
+    dims <- getSets(m)[c(1, 2)]
+    mainDim <- 1.1
+
+    for (a in tr) {
+      mTr <-   .histDim(m, tr = a, mainDim = mainDim, dims = dims)
+      # fill data
+      if (overwrite == TRUE) {
+        # only overwrite the exact combinations for which weighting existed, rest remains zero
+        addM <- setdiff(getItems(mTr, dim = 1), getItems(m, dim = 1))
+        if (length(addM) > 0) {
+          addMissM <- new.magpie(cells_and_regions = addM,
+                                 years = getItems(m, dim = 2),
+                                 names = getItems(m, dim = 3),
+                                 fill = 0)
+          m <- mbind(m, addMissM)
+        }
+
+        m[getItems(mTr, dim = 1), getYears(mTr), ] <- mTr
+      } else {
+        m[getItems(mTr, dim = 1), getYears(mTr), ][is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] <- mTr[is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] # nolint
+      }
+
     }
-  } # a in tr - transitions
 
-  # delete old lines
-  for (b in mapping$fromISO) {
-    if (is.element(b, getItems(m, dim = 1.1))) {
-      m <- m[-which(getItems(m, dim = 1.1) == b), , ]
+    # do second dimension
+    mainDim <- 1.2
+
+    for (a in tr2) {
+      mTr <-   .histDim(m, tr = a, mainDim = mainDim, dims = dims)
+      # fill data
+      if (overwrite == TRUE) {
+        # only overwrite the exact combinations for which weighting existed, rest remains zero
+        addM <- setdiff(getItems(mTr, dim = 1), getItems(m, dim = 1))
+        if (length(addM) > 0) {
+          addMissM <- new.magpie(cells_and_regions = addM,
+                                 years = getItems(m, dim = 2),
+                                 names = getItems(m, dim = 3),
+                                 fill = 0)
+          m <- mbind(m, addMissM)
+        }
+
+        m[getItems(mTr, dim = 1), getYears(mTr), ] <- mTr
+      } else {
+        m[getItems(mTr, dim = 1), getYears(mTr), ][is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] <- mTr[is.na(m[getItems(mTr, dim = 1), getYears(mTr), ])] # nolint
```

Comment:
I've seen that line before, can we get rid of this code duplication?

Reply:
i would rather opt to keep this one here... as it's the 2 separate loops doing filling in...